src/process_collector.rs

```diff
@@ -0,0 +1,346 @@
+
+        ProcessCollector {
+            pid: pid,
+            descs: descs,
+            cpu_total: Mutex::new(cpu_total),
+            open_fds: open_fds,
+            max_fds: max_fds,
+            vsize: vsize,
+            rss: rss,
+            start_time: start_time,
+        }
+    }
+
+    /// Returns a `ProcessCollector` of the calling process.
+    pub fn for_self() -> ProcessCollector {
+        let pid = unsafe { libc::getpid() };
+        ProcessCollector::new(pid, "")
+    }
+}
+
+impl Collector for ProcessCollector {
+    fn desc(&self) -> Vec<&Desc> {
+        self.descs.iter().collect()
+    }
+
+    fn collect(&self) -> Vec<proto::MetricFamily> {
+        // file descriptors
+        if let Ok(num) = open_fds(self.pid) {
+            self.open_fds.set(num as f64);
+        }
+        if let Ok(max) = max_fds(self.pid) {
+            self.max_fds.set(max)
+        }
+
+        // memory
+        if let Ok(statm) = pid_info::statm(self.pid) {
+            self.vsize.set(statm.size as f64 * *PAGESIZE);
+            self.rss.set(statm.resident as f64 * *PAGESIZE);
+        }
+
+        let pid_stat = pid_info::stat(self.pid);
+
+        // proc_start_time
+        if let (&Ok(ref stat), Some(boot_time)) = (&pid_stat, *BOOT_TIME) {
+            self.start_time.set(stat.start_time as f64 / *CLK_TCK + boot_time);
+        }
+
+        // cpu
+        let cpu_total_mfs = {
+            let cpu_total = self.cpu_total.lock().unwrap();
+            if let Ok(stat) = pid_stat {
+                let total = (stat.utime + stat.stime) as f64 / *CLK_TCK;
+                let past = cpu_total.get();
+                let delta = total - past;
+                if delta > 0.0 {
+                    cpu_total.inc_by(delta).unwrap();
+                }
+            }
+
+            cpu_total.collect()
+        };
+
+        // collect MetricFamilys.
+        let mut mfs = Vec::with_capacity(MERTICS_NUMBER);
+        mfs.extend(cpu_total_mfs);
+        mfs.extend(self.open_fds.collect());
+        mfs.extend(self.max_fds.collect());
+        mfs.extend(self.vsize.collect());
+        mfs.extend(self.rss.collect());
+        mfs.extend(self.start_time.collect());
+        mfs
+    }
+}
+
+fn open_fds(pid: pid_t) -> Result<usize> {
+    let path = format!("/proc/{}/fd", pid);
+    try!(fs::read_dir(path)).fold(Ok(0), |acc, i| {
+        let mut acc = try!(acc);
+        let ty = try!(try!(i).file_type());
+        if !ty.is_dir() {
+            acc += 1;
+        }
+
+        Ok(acc)
+    })
+}
+
+// `find_statistic` match lines in pattern pat, it takes the first matching line,
+// and parse the first number literal in the matching line.
+//
+// Example:
+//  * all:
+// ```
+// ctxt 789298306
+// btime 1477460662
+// processes 302136
+// procs_running 1
+// procs_blocked 0
+// ```
+//  * pat: "btime"
+//
+// then it returns `Ok(1477460662.0)`
+//
+// pub for tests.
+pub fn find_statistic(all: &str, pat: &str) -> Result<f64> {
```

Comment:
why not:

1. Split the line with space -> (name, value)
2. Compare the name directly
3. Return the value